File: simple-stack-example-mvp/src/main/java/com/zhuinden/simplestackdemoexamplemvp/util/BasePresenter.java
package com.zhuinden.simplestackdemoexamplemvp.util;

import android.view.View;

/**
 * Created by Owner on 2017. 01. 27..
 */

public abstract class BasePresenter<C extends BaseCoordinator<? extends View>> {
    C coordinator;

    public C getCoordinator() {
        return coordinator;
    }

    public final void attachCoordinator(C coordinator) {
        this.coordinator = coordinator;
        onAttach(coordinator);
    }

    public final void detachCoordinator(C coordinator) {
        onDetach(coordinator);
        this.coordinator = null;
    }

    protected abstract void onAttach(C coordinator);

    protected abstract void onDetach(C coordinator);
}

File: samples/legacy-architecture-samples/simple-stack-example-mvp-view/src/main/java/com/zhuinden/simplestackdemoexamplemvp/core/mvp/BasePresenter.kt
package com.zhuinden.simplestackdemoexamplemvp.core.mvp

import android.view.View

/**
 * Created by Owner on 2017. 01. 27..
 */

@Deprecated(message = "Manual view event dispatch is an anti-pattern.")
abstract class BasePresenter<V : View>: MvpPresenter<V> {
    var view: V? = null
        private set

    final override fun attachView(view: V) {
        this.view = view
        onAttach(view)
    }

    final override fun detachView(view: V) {
        onDetach(view)
        this.view = null
    }

    protected open fun onAttach(view: V) {}

    protected open fun onDetach(view: V) {}
}
